Polyhedron/demo/Polyhedron/Plugins/Alpha_wrap_3/Alpha_wrap_3_plugin.cpp

```diff
@@ -127,85 +169,143 @@ struct Iterative_AW3_visualization_visitor
   }
+};
 
-  template <typename AlphaWrapper, typename VertexHandle>
-  void after_Steiner_point_insertion(const AlphaWrapper&,
-                                     const VertexHandle) { }
+struct Wrapper_thread
+  : public QThread
+{
+  Q_OBJECT
 
-  template <typename AlphaWrapper>
-  void on_flood_fill_end(const AlphaWrapper&) { }
+  using Visitor = AW3_interrupter_visitor<Iterative_AW3_visualization_visitor>;
 
-  template <typename AlphaWrapper>
-  void on_alpha_wrapping_end(const AlphaWrapper&)
+public:
+  Wrapper wrapper;
+  const Oracle oracle;
+  const double alpha, offset;
+  const bool enforce_manifoldness;
+  Visitor visitor;
+
+  SMesh wrap;
+
+  QTimer* timer;
+
+public:
+  Wrapper_thread(const Oracle& oracle,
+                 const double alpha,
+                 const double offset,
+                 const bool enforce_manifoldness,
+                 Visitor visitor)
+    : wrapper(oracle),
+      alpha(alpha), offset(offset),
+      enforce_manifoldness(enforce_manifoldness),
+      visitor(visitor),
+      timer(new QTimer(this))
   {
-    if(m_iterative_wrap_item == nullptr)
-      return;
+    connect(timer, SIGNAL(timeout()),
+            this,  SLOT(emit_status()));
 
-    m_iterative_wrap_item->setName(QString("Iterative wrap #%1").arg(sid));
+    timer->start(1000);
+  }
 
-    m_iterative_wrap_item->setAlpha(255);
-    m_iterative_wrap_item->invalidateOpenGLBuffers();
-    m_iterative_wrap_item->redraw();
-    m_iterative_wrap_item->itemChanged();
+  ~Wrapper_thread()
+  {
+    delete timer;
+  }
 
-    QApplication::processEvents();
+  void run() override
+  {
+    QElapsedTimer elapsed_timer;
+    elapsed_timer.start();
 
-    if(m_do_snapshot)
+    // try-catch because the stop visitor currently uses a throw
+    try
     {
-      std::stringstream oss;
-      oss << "Wrap_iteration-" << sid << ".png" << std::ends;
-      QString filename = QString::fromStdString(oss.str().c_str());
+      wrapper(alpha, offset, wrap,
+              CGAL::parameters::do_enforce_manifoldness(enforce_manifoldness)
+                               .visitor(visitor));
 
-      CGAL::Three::Viewer_interface* viewer = CGAL::Three::Three::activeViewer();
-      viewer->saveSnapshot(filename);
+      Q_EMIT done(this);
     }
+    catch(const Out_of_patience_exception&)
+    {
+      if(enforce_manifoldness)
+        wrapper.make_manifold();
 
-    m_iterative_wrap_item->setVisible(false);
+      // extract the wrap in its current state
+      wrapper.extract_surface(wrap, CGAL::get(CGAL::vertex_point, wrap), !enforce_manifoldness);
 
-    // Refresh the view
-    QApplication::processEvents();
+      Q_EMIT interrupted(this);
+    }
+
+    std::cout << "Wrapping took " << elapsed_timer.elapsed() / 1000. << "s" << std::endl;
   }
+
+public Q_SLOTS:
+  void emit_status()
+  {
+    Q_EMIT status_report(QString("%1 vertices").arg(wrapper.triangulation().number_of_vertices()));
+  }
+
```

Comment:
That is the only place where I might see an issue. But the Mesh_3 plugin has the same: the timer is created and started in the main thread, so this slot will be called by the main thread. The call `wrapper.triangulation().number_of_vertices()` is not thread-safe. However, the only consequence I can see is a possible incoherent number of vertices during one iteration of the slot.

Reply:
That is related to the macro `CGAL_CONCURRENT_COMPACT_CONTAINER_APPROXIMATE_SIZE` from PR #4389.